File: Python/ByRound/1672/1672_F1_Array_Shuffling.py
# Problem: CF 1672 F1 - Array Shuffling
# https://codeforces.com/contest/1672/problem/F1
 
"""
Algorithm: Array Shuffling for Maximum Sadness
Purpose: Given an array 'a', find a permutation 'b' such that the minimum number of swaps to transform 'b' into 'a' is maximized.
Approach:
1. Group indices of elements by their value to form cycles.
2. For each group of indices with the same value, sort them based on the values they point to.
3. Assign new values to these indices such that the resulting array 'b' has maximum sadness.
Time Complexity: O(n log n) due to sorting within cycles
Space Complexity: O(n) for storing counts, cycles, and the new array
"""
 
import sys
 
I = lambda: [*map(int, sys.stdin.readline().split())]  # Fast input helper
 
(t,) = I()
for _ in range(t):
    (n,) = I()
    a = I()
    counts = [0] * (n + 1)  # Count of occurrences of each value
    cycles = []  # List to store indices grouped by value
    
    # Group indices by their values
    for i in range(n):
        val = a[i]
        ind = counts[val]  # Determine which cycle this index belongs to
        if len(cycles) <= ind:
            cycles.append([i])  # Create new cycle if needed
        else:
            cycles[ind].append(i)  # Add index to existing cycle
        counts[val] += 1  # Increment count for this value
 
    new = [0] * n  # Resulting array
    # For each cycle, rearrange indices to maximize swaps needed
    for cycle in cycles:
        cycle.sort(key=lambda x: a[x])  # Sort cycle indices by the values at those indices
        l = len(cycle)
        for i in range(l):
            new[cycle[i - 1]] = a[cycle[i]]  # Assign value from next index in cycle (circular)
    print(*new)
 
 
# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/F/1672_F1_Array_Shuffling.cpp
// Problem: CF 1672 F1 - Array Shuffling
// https://codeforces.com/contest/1672/problem/F1
 
/*
 * Problem: F1. Array Shuffling
 * Purpose: Given an array 'a', find a permutation 'b' such that the minimum
 * number of swaps needed to transform 'b' back to 'a' (called sadness) is
 * maximized.
 *
 * Algorithm:
 *   - The sadness of an array is the number of cycles in the permutation that
 * transforms it into 'a'.
 *   - To maximize the sadness, we want to create as many cycles as possible.
 *   - Strategy:
 *     1. Use a multiset to keep track of available elements.
 *     2. For each position, choose the smallest element greater than the
 * current element in the multiset. If no such element exists, choose the
 * smallest one.
 *     3. Remove used elements from the multiset to avoid reuse.
 *
 * Time Complexity: O(n log n) per test case due to multiset operations (insert,
 * erase, upper_bound). Space Complexity: O(n) for storing the array and the
 * multiset.
 */
 
#include <cstdio>
#include <iostream>
#include <set>
 
using namespace std;
int t, n, a[200001];
int main() {
  cin >> t;
  for (int i = 1; i <= t; i++) {
    multiset<int> s;  // Multiset to maintain sorted elements and support
                      // efficient operations
    cin >> n;
    for (int i = 1; i <= n; i++) {
      cin >> a[i];
      s.insert(a[i]);  // Read array and insert into multiset
    }
    for (int i = 1; i <= n; i++) {
      multiset<int>::iterator it =
          s.upper_bound(a[i]);            // Find smallest element > a[i]
      if (it == s.end()) it = s.begin();  // If none exists, pick the smallest
      cout << *it << " ";                 // Output the chosen element
      s.erase(it);                        // Remove chosen element from multiset
    }
    cout << endl;
  }
}
 
// https://github.com/VaHiX/CodeForces/